File: src/main/java/org/apache/sysml/hops/spoof2/Spoof2Compiler.java
package org.apache.sysml.hops.spoof2;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.apache.sysml.hops.AggBinaryOp;
import org.apache.sysml.hops.AggUnaryOp;
import org.apache.sysml.hops.BinaryOp;
import org.apache.sysml.hops.DataGenOp;
import org.apache.sysml.hops.DataOp;
import org.apache.sysml.hops.Hop;
import org.apache.sysml.hops.Hop.AggOp;
import org.apache.sysml.hops.Hop.Direction;
import org.apache.sysml.hops.Hop.OpOp1;
import org.apache.sysml.hops.Hop.OpOp2;
import org.apache.sysml.hops.Hop.OpOp3;
import org.apache.sysml.hops.HopsException;
import org.apache.sysml.hops.LiteralOp;
import org.apache.sysml.hops.ReorgOp;
import org.apache.sysml.hops.UnaryOp;
import org.apache.sysml.hops.rewrite.HopRewriteUtils;
import org.apache.sysml.hops.rewrite.ProgramRewriteStatus;
import org.apache.sysml.hops.rewrite.ProgramRewriter;
import org.apache.sysml.hops.spoof2.plan.SNode;
import org.apache.sysml.hops.spoof2.plan.SNodeAggregate;
import org.apache.sysml.hops.spoof2.plan.SNodeData;
import org.apache.sysml.hops.spoof2.plan.SNodeExt;
import org.apache.sysml.hops.spoof2.plan.SNodeNary;
import org.apache.sysml.hops.spoof2.plan.SNodeNary.JoinCondition;
import org.apache.sysml.hops.spoof2.plan.SNodeNary.NaryOp;
import org.apache.sysml.hops.spoof2.rewrite.BasicSPlanRewriter;
import org.apache.sysml.hops.spoof2.rewrite.SNodeRewriteUtils;
import org.apache.sysml.parser.DMLProgram;
import org.apache.sysml.parser.ForStatement;
import org.apache.sysml.parser.ForStatementBlock;
import org.apache.sysml.parser.FunctionStatement;
import org.apache.sysml.parser.FunctionStatementBlock;
import org.apache.sysml.parser.IfStatement;
import org.apache.sysml.parser.IfStatementBlock;
import org.apache.sysml.parser.LanguageException;
import org.apache.sysml.parser.StatementBlock;
import org.apache.sysml.parser.WhileStatement;
import org.apache.sysml.parser.WhileStatementBlock;
import org.apache.sysml.runtime.DMLRuntimeException;
import org.apache.sysml.utils.Explain;
import org.spark_project.guava.collect.Lists;

public class Spoof2Compiler 
{
	private static final Log LOG = LogFactory.getLog(Spoof2Compiler.class.getName());
	
	//internal configuration flags
	public static boolean LDEBUG = true;
	
	static {
		// for internal debugging only
		if( LDEBUG ) {
			Logger.getLogger("org.apache.sysml.hops.spoof2")
				  .setLevel(Level.TRACE);
		}
	}
	
	
	public static void generateCode(DMLProgram dmlprog) 
		throws LanguageException, HopsException, DMLRuntimeException
	{
		// for each namespace, handle function statement blocks
		for (String namespaceKey : dmlprog.getNamespaces().keySet()) {
			for (String fname : dmlprog.getFunctionStatementBlocks(namespaceKey).keySet()) {
				FunctionStatementBlock fsblock = dmlprog.getFunctionStatementBlock(namespaceKey,fname);
				generateCodeFromStatementBlock(fsblock);
			}
		}
		
		// handle regular statement blocks in "main" method
		for (int i = 0; i < dmlprog.getNumStatementBlocks(); i++) {
			StatementBlock current = dmlprog.getStatementBlock(i);
			generateCodeFromStatementBlock(current);
		}
	}
	
	public static void generateCodeFromStatementBlock(StatementBlock current)
		throws HopsException, DMLRuntimeException
	{
		if (current instanceof FunctionStatementBlock)
		{
			FunctionStatementBlock fsb = (FunctionStatementBlock)current;
			FunctionStatement fstmt = (FunctionStatement)fsb.getStatement(0);
			for (StatementBlock sb : fstmt.getBody())
				generateCodeFromStatementBlock(sb);
		}
		else if (current instanceof WhileStatementBlock)
		{
			WhileStatementBlock wsb = (WhileStatementBlock) current;
			WhileStatement wstmt = (WhileStatement)wsb.getStatement(0);
			wsb.setPredicateHops(optimize(wsb.getPredicateHops(), false));
			for (StatementBlock sb : wstmt.getBody())
				generateCodeFromStatementBlock(sb);
		}
		else if (current instanceof IfStatementBlock)
		{
			IfStatementBlock isb = (IfStatementBlock) current;
			IfStatement istmt = (IfStatement)isb.getStatement(0);
			isb.setPredicateHops(optimize(isb.getPredicateHops(), false));
			for (StatementBlock sb : istmt.getIfBody())
				generateCodeFromStatementBlock(sb);
			for (StatementBlock sb : istmt.getElseBody())
				generateCodeFromStatementBlock(sb);
		}
		else if (current instanceof ForStatementBlock) //incl parfor
		{
			ForStatementBlock fsb = (ForStatementBlock) current;
			ForStatement fstmt = (ForStatement)fsb.getStatement(0);
			fsb.setFromHops(optimize(fsb.getFromHops(), false));
			fsb.setToHops(optimize(fsb.getToHops(), false));
			fsb.setIncrementHops(optimize(fsb.getIncrementHops(), false));
			for (StatementBlock sb : fstmt.getBody())
				generateCodeFromStatementBlock(sb);
		}
		else //generic (last-level)
		{
			current.set_hops( generateCodeFromHopDAGs(current.get_hops()) );
			current.updateRecompilationFlag();
		}
	}

	public static ArrayList<Hop> generateCodeFromHopDAGs(ArrayList<Hop> roots) 
		throws HopsException, DMLRuntimeException
	{
		if( roots == null )
			return null;

		ArrayList<Hop> optimized = optimize(roots, false);
		Hop.resetVisitStatus(roots);
		Hop.resetVisitStatus(optimized);
		
		return optimized;
	}
	
	
	/**
	 * Main interface of sum-product optimizer, predicate dag.
	 * 
	 * @param root dag root node
	 * @param recompile true if invoked during dynamic recompilation
	 * @return dag root node of modified dag
	 * @throws DMLRuntimeException if optimization failed
	 * @throws HopsException -
	 */
	public static Hop optimize( Hop root, boolean recompile ) throws DMLRuntimeException, HopsException {
		if( root == null )
			return null;
		return optimize(new ArrayList<Hop>(Arrays.asList(root)), recompile).get(0);
	}
	
	/**
	 * Main interface of sum-product optimizer, statement block dag.
	 * 
	 * @param roots dag root nodes
	 * @param recompile true if invoked during dynamic recompilation
	 * @return dag root nodes of modified dag 
	 * @throws DMLRuntimeException if optimization failed
	 * @throws HopsException -
	 */
	public static ArrayList<Hop> optimize(ArrayList<Hop> roots, boolean recompile) 
		throws DMLRuntimeException, HopsException 
	{
		if( LOG.isTraceEnabled() ) {
			LOG.trace("Spoof2Compiler called for HOP DAG: \n" 
				+ Explain.explainHops(roots));
		}
		
		//construct sum-product plan
		ArrayList<SNode> sroots = new ArrayList<>();
		for( Hop root : roots )
			sroots.add(rConstructSumProductPlan(root));
		
		if( LOG.isTraceEnabled() ) {
			LOG.trace("Explain after initial SPlan construction: " 
				+ Explain.explainSPlan(sroots));
		}
		
		//rewrite sum-product plan
		BasicSPlanRewriter rewriter = new BasicSPlanRewriter();
		sroots = rewriter.rewriteSPlan(sroots);
		
		if( LOG.isTraceEnabled() ) {
			LOG.trace("Explain after SPlan rewriting: " 
				+ Explain.explainSPlan(sroots));
		}
		
		//re-construct modified HOP DAG
		ArrayList<Hop> roots2 = new ArrayList<>();
		for( SNode sroot : sroots )
			roots2.add(rReconstructHopDag(sroot));
		
		if( LOG.isTraceEnabled() ) {
			LOG.trace("Spoof2Compiler created modified HOP DAG: \n" 
				+ Explain.explainHops(roots2));
		}
		
		//rewrite after applied sum-product optimizer
		Hop.resetVisitStatus(roots2);
		ProgramRewriter rewriter2 = new ProgramRewriter(true, true);
		roots2 = rewriter2.rewriteHopDAGs(roots2, new ProgramRewriteStatus());
	
		if( LOG.isTraceEnabled() ) {
			LOG.trace("Spoof2Compiler rewritten modified HOP DAG: \n" 
				+ Explain.explainHops(roots2));
		}
		
		return roots2;
	}
	
	private static SNode rConstructSumProductPlan(Hop current) {
		//recursively process child nodes
		ArrayList<SNode> inputs = new ArrayList<>();
		for( Hop c : current.getInput() )
			inputs.add(rConstructSumProductPlan(c));
		
		//construct node for current hop
		SNode node = null;
		if( current instanceof DataOp ) {
			node = ((DataOp)current).isWrite() ?
				new SNodeData(inputs.get(0), current) : 
				new SNodeData(current);
		}
		else if( current instanceof LiteralOp ) {
			node = new SNodeData(current);
		}
		else if( current instanceof DataGenOp ) {
			node = new SNodeExt(current);
		}
		else if( current instanceof ReorgOp ) {
			node = HopRewriteUtils.isTransposeOperation(current) ?
				new SNodeNary(inputs.get(0), NaryOp.TRANSPOSE,
					Lists.reverse(inputs.get(0).getSchema())):
				new SNodeExt(inputs.get(0), current);
		}
		else if( current instanceof UnaryOp ) {
			OpOp1 op = ((UnaryOp) current).getOp();
			node = NaryOp.contains(op.name()) ?
				new SNodeNary(inputs.get(0), NaryOp.valueOf(op.name())) :
				new SNodeExt(inputs.get(0), current);
		}
		else if( current instanceof BinaryOp ) {
			OpOp2 op = ((BinaryOp) current).getOp();
			node = NaryOp.contains(op.name()) ?
				new SNodeNary(inputs, NaryOp.valueOf(op.name())) :
				new SNodeExt(inputs, current);			
		}
		else if( current instanceof AggUnaryOp ) {
			AggUnaryOp au = (AggUnaryOp) current;
			List<String> schema = (au.getDirection()==Direction.Row) ? 
				Arrays.asList(inputs.get(0).getSchema().get(0)) : (au.getDirection()==Direction.Col) ? 
				Arrays.asList(inputs.get(0).getSchema().get(1)) : new ArrayList<String>();
			node = new SNodeAggregate(inputs.get(0), 
				((AggUnaryOp)current).getOp(), schema);
		}
		else if( current instanceof AggBinaryOp ) {
			SNode mult = new SNodeNary(inputs, NaryOp.MULT, 
				new JoinCondition(inputs.get(0).getSchema().get(1),
				inputs.get(1).getSchema().get(0)));
			mult.setDims(current.getDim1(), current.getDim2(), 
				current.getInput().get(0).getDim2());
			node = new SNodeAggregate(mult, AggOp.SUM, Arrays.asList(
				mult.getSchema().get(0), mult.getSchema().get(2)));
		}
		
		//check for valid created SNode
		if( node == null ) {
			throw new RuntimeException("Error constructing SNode for HOP: " +
				current.getHopID() + " " + current.getOpString() + ".");
		}
		
		//update size information (other than intermediates)
		node.setDims(current.getDim1(), current.getDim2());
		
		return node;
	}
	
	private static Hop rReconstructHopDag(SNode current) {
		
		//recursively process child nodes
		ArrayList<Hop> inputs = new ArrayList<Hop>();
		for( SNode c : current.getInput() )
			inputs.add(rReconstructHopDag(c));
		
		Hop node = null;
		
		if( current instanceof SNodeData ) {
			node = ((SNodeData)current).getHop();
			if( !current.getInput().isEmpty() ) {
				HopRewriteUtils.replaceChildReference(node, 
					node.getInput().get(0), inputs.get(0), 0);
			}
		}
		else if( current instanceof SNodeAggregate ) {
			SNodeAggregate agg = (SNodeAggregate) current;
			if( SNodeRewriteUtils.isSNodeNary(agg.getInput().get(0), NaryOp.MULT) 
				&& (agg.isScalar() || agg.getInput().get(0).getNumDims()==3) ) {
				//TODO proper handling of schemas to decide upon transpose
				Hop input1 = inputs.get(0).getInput().get(0);
				Hop input2 = inputs.get(0).getInput().get(1);
				input1 = (agg.isScalar() && input1.getDim1() > 1) ? 
					HopRewriteUtils.createTranspose(input1) : input1;
				input2 = (agg.isScalar() && input2.getDim2() > 1) ? 
						HopRewriteUtils.createTranspose(input2) : input2;
				node = HopRewriteUtils.createMatrixMultiply(input1, input2);
				node = HopRewriteUtils.createUnary(node, OpOp1.CAST_AS_SCALAR);
			}
			else {
				Direction dir = agg.getSchema().isEmpty() ? Direction.RowCol : 
					agg.getSchema().get(0).equals(agg.getInput().get(0).getSchema().get(0)) ? 
					Direction.Row : Direction.Col;
				node = HopRewriteUtils.createAggUnaryOp(inputs.get(0), agg.getOp(), dir);
			}
		}
		else if( current instanceof SNodeNary ) {
			SNodeNary nary = (SNodeNary) current;
			if( inputs.size()==1 ) { //unary
				if( nary.getOp()==NaryOp.TRANSPOSE )
					node = HopRewriteUtils.createTranspose(inputs.get(0));
				else
					node = HopRewriteUtils.createUnary(
						inputs.get(0), OpOp1.valueOf(nary.getOp().name()));
			}
			else if( inputs.size()==2 ) { //binary
				node = HopRewriteUtils.createBinary(inputs.get(0), 
					inputs.get(1), OpOp2.valueOf(nary.getOp().name()));
			}
			else if( inputs.size()==3 ) { //ternary
				node = HopRewriteUtils.createTernary(inputs.get(0), inputs.get(1), 
					inputs.get(2), OpOp3.valueOf(nary.getOp().name()));
			}
		}
		else if( current instanceof SNodeExt ) {
			node = ((SNodeExt)current).getHop();
			if( !inputs.isEmpty() ) { //robustness datagen
				HopRewriteUtils.removeAllChildReferences(node);
				for( Hop c : inputs )
					node.addInput(c);
			}
		}
		
		//check for valid created SNode
		if( node == null ) {
			throw new RuntimeException("Error constructing Hop for SNode: " +
				current.getId() + " " + current.toString() + ".");
		}
		
		return node;
	} 
}

File: src/main/java/org/apache/sysml/hops/spoof2/Spoof2Compiler.kt
package org.apache.sysml.hops.spoof2

import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.ListMultimap
import org.apache.commons.logging.LogFactory
import org.apache.log4j.Level
import org.apache.log4j.Logger
import org.apache.sysml.conf.DMLConfig
import org.apache.sysml.hops.*
import org.apache.sysml.hops.rewrite.*
import org.apache.sysml.hops.spoof2.enu.NormalFormExploreEq
import org.apache.sysml.hops.spoof2.plan.*
import org.apache.sysml.hops.spoof2.rewrite.SPlanRewriteRule
import org.apache.sysml.hops.spoof2.rewrite.SPlanRewriter
import org.apache.sysml.parser.*
import org.apache.sysml.runtime.DMLRuntimeException
import org.apache.sysml.utils.Explain
import java.util.*

object Spoof2Compiler {
    internal val LOG = LogFactory.getLog(Spoof2Compiler::class.java.name)

    //internal configuration flags
    private const val LDEBUG = DMLConfig.SPOOF_DEBUG
    init {
        if (LDEBUG) Logger.getLogger("org.apache.sysml.hops.spoof2").level = Level.TRACE
    }

    // todo inefficient; might be fine just to check the top-level sizes
    fun allKnownSizes(hop: Hop): Boolean {
        if (!hop.dimsKnown())
            return false
        return hop.input.all { allKnownSizes(it) }
    }
    fun allKnownSizesAndNoPermuteMultiply(hop: Hop): Boolean {
        if (!hop.dimsKnown() || hop is AggBinaryOp && hop.hasLeftPMInput())
            return false
        return hop.input.all { allKnownSizesAndNoPermuteMultiply(it) }
    }

    @JvmStatic
    @Throws(LanguageException::class, HopsException::class, DMLRuntimeException::class)
    fun generateCode(dmlprog: DMLProgram) {
        // for each namespace, handle function statement blocks
        for (namespaceKey in dmlprog.namespaces.keys) {
            for (fname in dmlprog.getFunctionStatementBlocks(namespaceKey).keys) {
                val fsblock = dmlprog.getFunctionStatementBlock(namespaceKey, fname)
                generateCodeFromStatementBlock(fsblock)
            }
        }

        // handle regular statement blocks in "main" method
        for (i in 0 until dmlprog.numStatementBlocks) {
            val current = dmlprog.getStatementBlock(i)
            generateCodeFromStatementBlock(current)
        }
    }

    @JvmStatic
    @Throws(HopsException::class, DMLRuntimeException::class)
    fun generateCodeFromStatementBlock(current: StatementBlock) {
        when( current ) {
            is FunctionStatementBlock -> {
                val fstmt = current.getStatement(0) as FunctionStatement
                for (sb in fstmt.body)
                    generateCodeFromStatementBlock(sb)
            }
            is WhileStatementBlock -> {
                val wsb = current
                val wstmt = wsb.getStatement(0) as WhileStatement
                wsb.predicateHops = optimize(wsb.predicateHops, false, true)
                for (sb in wstmt.body)
                    generateCodeFromStatementBlock(sb)
            }
            is IfStatementBlock -> {
                val isb = current
                val istmt = isb.getStatement(0) as IfStatement
                isb.predicateHops = optimize(isb.predicateHops, false, true)
                for (sb in istmt.ifBody)
                    generateCodeFromStatementBlock(sb)
                for (sb in istmt.elseBody)
                    generateCodeFromStatementBlock(sb)
            }
            is ForStatementBlock -> { //incl parfor
                val fsb = current
                val fstmt = fsb.getStatement(0) as ForStatement
                fsb.fromHops = optimize(fsb.fromHops, false, true)
                fsb.toHops = optimize(fsb.toHops, false, true)
                fsb.incrementHops = optimize(fsb.incrementHops, false, true)
                for (sb in fstmt.body)
                    generateCodeFromStatementBlock(sb)
            }
            else -> { //generic (last-level)
                current._hops = generateCodeFromHopDAGs(current._hops, false, true)
                current.updateRecompilationFlag()
            }
        }
    }

    @JvmStatic
    @Throws(HopsException::class, DMLRuntimeException::class)
    fun generateCodeFromHopDAGs(roots: ArrayList<Hop>?, recompile: Boolean, doDynamicProgramRewriter: Boolean): ArrayList<Hop>? {
        if (roots == null)
            return null

        val optimized = optimize(roots, recompile, doDynamicProgramRewriter)
//        Hop.resetVisitStatus(roots)
        Hop.resetVisitStatus(optimized)

        return optimized
    }

    /**
     * Main interface of sum-product optimizer, predicate dag.

     * @param root      dag root node
     * @param recompile true if invoked during dynamic recompilation
     * @return dag root node of modified dag
     * @throws DMLRuntimeException if optimization failed
     * @throws HopsException       -
     */
    @JvmStatic
    @Throws(DMLRuntimeException::class, HopsException::class)
    fun optimize(root: Hop?, recompile: Boolean, doDynamicProgramRewriter: Boolean): Hop? {
        if (root == null)
            return null
        return optimize(ArrayList(Arrays.asList(root)), recompile, doDynamicProgramRewriter).get(0)
    }

    private fun programRewriteHops(roots: ArrayList<Hop>, recompile: Boolean,
                                   doDynamicProgramRewriter: Boolean): ArrayList<Hop> {
        Hop.resetVisitStatus(roots)
//        initial compile          ==>  CSE; SumProduct; static + dynamic ProgramWriter
//        recompile, inplace=false ==>  CSE; SumProduct; dynamic ProgramWriter
//        recompile, inplace=true  ==>  CSE; SumProduct
//        LOG.trace("Call ProgramRewriter with static=${!recompile} dynamic=$doDynamicProgramRewriter")
        val rewriter2 = ProgramRewriter(!recompile, doDynamicProgramRewriter)
        rewriter2.rewriteHopDAG(roots, ProgramRewriteStatus())
        Hop.resetVisitStatus(roots)
        return roots
    }

    // dim -> <Input -> Dim>
    private fun dimsToInputDims(root: SNode): List<ListMultimap<Hop, AU>> {
        return root.schema.map { (n,_) ->
            val mm = ArrayListMultimap.create<Hop, AU>()
            when( n ) {
                is AB -> mapToInputs(root, n, mm)
                is AU -> mapToInputs(root, n, mm)
            }
            mm
        }
    }

    private fun mapToInputs(node: SNode, i: AU, mm: ListMultimap<Hop, AU>) {
        when( node ) {
            is SNodeData -> {
                if( node.isWrite ) mapToInputs(node.inputs[0], i, mm)
                else if( node.isLiteral )
                else mm.put(node.hop, i)
            }
            is SNodeBind -> mapToInputs(node.input, i, mm)
            is SNodeUnbind -> {
                if( node.unbindings.containsKey(i) )
                    mapToInputs(node.input, node.unbindings[i]!!, mm)
                else
                    mapToInputs(node.input, i, mm)
            }
            is SNodeExt -> mm.put(node.hop, i)
            else -> throw SNodeException(node, "don't know how to handle snode type $node")
        }
    }
    private fun mapToInputs(node: SNode, n: AB, mm: ListMultimap<Hop, AU>) {
        when( node ) {
            is SNodeBind -> {
                if( n in node.bindings.values )
                    mapToInputs(node.input, node.bindings.entries.find { it.value == n }!!.key, mm)
                else
                    mapToInputs(node.input, n, mm)
            }
            else -> node.inputs.filter { n in it.schema }.map { mapToInputs(it, n, mm) }
        }
    }

    /**
     * Main interface of sum-product optimizer, statement block dag.

     * @param roots dag root nodes
     * @param recompile true if invoked during dynamic recompilation
     * @return dag root nodes of modified dag
     * @throws DMLRuntimeException if optimization failed
     * @throws HopsException -
     */
    @Throws(DMLRuntimeException::class, HopsException::class)
    fun optimize(roots: ArrayList<Hop>, recompile: Boolean, doDynamicProgramRewriter: Boolean): ArrayList<Hop> {

        if( !doDynamicProgramRewriter && roots.any { it !is DataOp } ) {
            // Todo: better handling of dynamic recompile during inplace. What is safe to change during recompile?
            // GLMDML fails to converge if this is not handled carefully.
            if( LOG.isTraceEnabled )
                LOG.trace("Skipping due to inplace recompile setting")
            return roots
        }

        // if any sizes unknown, don't do Spoof2
        if( roots.any { !allKnownSizesAndNoPermuteMultiply(it)} ) {
            if (LOG.isTraceEnabled)
                LOG.trace("Skipping Spoof2 due to unknown sizes or a permutation MxM")
            return programRewriteHops(roots, recompile, doDynamicProgramRewriter)
        }

        ProgramRewriter(RewriteCommonSubexpressionElimination()).rewriteHopDAG(roots, ProgramRewriteStatus())

        if (LOG.isTraceEnabled)
            LOG.trace("Spoof2Compiler called for HOP DAG${if(recompile) " at RECOMPILE" else ""}: \n" + Explain.explainHops(roots))

        // remember top-level orientations
        val rootClasses = roots.map(Hop::classify)

        HopDagValidator.validateHopDag(roots)

        //construct sum-product plan
        var sroots = Hop2SPlan.hop2SPlan(roots)

        // if this is entirely composed of Data and Ext ops, then don't do Spoof2 because nothing to do
        if( sroots.all { it.isEntirelyDataExtEquals() } ) {
            if (LOG.isTraceEnabled) {
                LOG.trace("Skipping Spoof2 on DAG that is entirely composed of Data, Ext, and == nodes")
            }
            return programRewriteHops(roots, recompile, doDynamicProgramRewriter)
        }

        // for each root, a map from input dimensions (actually a list) to the base inputs and input dimensions
        val baseInputs = sroots.map { sroot -> dimsToInputDims(sroot) }

//        if (LOG.isTraceEnabled)
//            LOG.trace("Explain after initial SPlan construction: " + Explain.explainSPlan(sroots))

        val result2NormalForm = SPlan2NormalForm_InsertStyle.rewriteSPlan(sroots) // SPlan2NormalForm.rewriteSPlan(sroots)
        sroots = result2NormalForm.replace(sroots)

        if( result2NormalForm != SPlanRewriter.RewriterResult.NoChange ) {
            sroots = NormalFormExploreEq().rewriteSPlan(sroots).replace(sroots)
            if( SPlanRewriteRule.LOG.isTraceEnabled )
                SPlanRewriteRule.LOG.trace("After processing normal form: "+Explain.explainSPlan(sroots))
        }

        //re-construct modified HOP DAG
//        sroots = SPlan2HopReady.rewriteSPlan(sroots).replace(sroots) // put in SPlan2Hop
        var roots2 = SPlan2Hop.splan2Hop(sroots)

        val baseInputs2 = sroots.map { sroot ->
            dimsToInputDims(sroot)
        }
        if( LOG.isTraceEnabled ) {
            LOG.trace("Base input map: ${baseInputs.mapIndexed {i, bil -> "\nr$i\t${bil.mapIndexed { d, bi -> "d$d: ${bi.entries().groupBy { it.key }.map { "(${it.key.hopID})${it.key.opString}=${it.value.map { it.value }}" }.joinToString()}"}.joinToString("\n\t")}"}}" +
                      "\nBase input map2: ${baseInputs2.mapIndexed {i, bil -> "\nr$i\t${bil.mapIndexed { d, bi -> "d$d: ${bi.entries().groupBy { it.key }.map { "(${it.key.hopID})${it.key.opString}=${it.value.map { it.value }}" }.joinToString()}"}.joinToString("\n\t")}"}}")
        }


        // add transposes if necessary to roots in order to maintain same orientation as original
        // shouldn't be necessary because the roots are generally Writes, which correct orientation on their own
        roots2.mapInPlaceIndexed { idx, root2 ->
            if( rootClasses[idx].isVector && root2.classify() != rootClasses[idx] ) {
                check( root2.classify().isVector ) {"root changed type after reconstruction? Old type ${rootClasses[idx]}; new type ${root2.classify()} dims ${root2.dim1}, ${root2.dim2} hopid=${root2.hopID}" +
                        "\n modified Hop Dag is:\n" + Explain.explainHops(roots2)}
                if( LOG.isTraceEnabled )
                    LOG.trace("creating root transpose at root $idx to enforce orientation ${rootClasses[idx]}")
                transposeRoot(root2)
            } else if( rootClasses[idx] == HopClass.MATRIX ) { // check data flow
                val bi = baseInputs[idx]
                // does dim 0 base inputs == dim 1 base inputs? If so, it is symmetric and orientation does not matter.
                if( bi[0] == bi[1] )
                    root2
                else {
                    val bi2 = baseInputs2[idx]
                    // Does bi[0] and bi[1] have different unique inputs? If so, distinguish based on # of unique inputs.
                    val biu = bi.map { it.asMap().mapValues { (_,v) -> v.distinct() }.toList().distinct() }
                    if( biu[0] != biu[1] ) {
                        val biu2 = bi2.map { it.asMap().mapValues { (_,v) -> v.distinct() }.toList().distinct() }
                        when {
                            biu[0] == biu2[0] -> // need better method?
                                root2
                            biu[0] == biu2[1] -> {
                                if( LOG.isTraceEnabled )
                                    LOG.trace("create root transpose at root $idx based on data flow; unique inputs differ case")
                                transposeRoot(root2)
                            }
                            biu[0].toSet() == biu2[0].toSet() -> root2
                            biu[0].toSet() == biu2[1].toSet() -> {
                                if( LOG.isTraceEnabled )
                                    LOG.trace("create root transpose at root $idx based on data flow; unique inputs differ case (by set)")
                                transposeRoot(root2)
                            }
                            else -> throw SNodeException(sroots[idx], "Failed to distinguish orientation for old Hop root ${roots[idx].hopID} and new Hop root ${root2.hopID}; " +
                                    "baseInputs have different unique inputs but neither equal to base inputs 2:" +
                                    "\n\t0: ${biu[0].map { (k,_) -> k }.toSet().joinToString { "(${it.hopID})${it.opString}" }}" +
                                    "\n\t1: ${biu[1].map { (k,_) -> k }.toSet().joinToString { "(${it.hopID})${it.opString}" }}" +
                                    "\n\t0: ${biu2[0].map { (k,_) -> k }.toSet().joinToString { "(${it.hopID})${it.opString}" }}" +
                                    "\n\t1: ${biu2[0].map { (k,_) -> k }.toSet().joinToString { "(${it.hopID})${it.opString}" }}")
                        }
                    } else {
                        // for each unique base input, if the number of base inputs is different between dimensions,
                        // go with the orientation that agrees with the disparity
                        @Suppress("NAME_SHADOWING")
                        val r = bi[0].keySet().fold(-1) { commit, uniqueBase ->
                            // bi[0] is the base input of root dimension 0
                            // x1[0] corresponds to bi[0]
                            // x1[0][0] is the first dimension of the base input
                            val x1 = bi.map { it[uniqueBase]!!.groupBy { it }.map { it.key to it.value.count() }.sortedBy { it.first }.map { it.second }}
                            val x2 = bi2.map { it[uniqueBase]!!.groupBy { it }.map { it.key to it.value.count() }.sortedBy { it.first }.map { it.second }}
                            // -1 -> uncommitted; 0 -> commit no transpose; 1 -> commit transpose
                            x1[0].indices.fold(commit) { commit, baseIdx ->
                                val c = x1[0][baseIdx].compareTo(x1[1][baseIdx])
                                if( c == 0 )
                                    commit
                                else {
                                    val c2 = x2[0][baseIdx].compareTo(x2[1][baseIdx])
                                    val newCommit = when (c2) {
                                        0 -> commit
                                        c -> 0
                                        else -> 1
                                    }
                                    if( commit >= 0 && commit != newCommit )
                                        throw SNodeException(sroots[idx], "Failed to distinguish orientation for old Hop root ${roots[idx].hopID} and new Hop root ${root2.hopID}; " +
                                                "baseInputs have same unique inputs and on input $uniqueBase, diagreeement with a previous unique input")
                                    newCommit
                                }
                            }
                        }
                        when( r ) {
                            0 -> root2
                            1 -> {
                                if( LOG.isTraceEnabled )
                                    LOG.trace("create root transpose at root $idx based on data flow; unique inputs equal case")
                                transposeRoot(root2)
                            }
                            -1 -> root2 // symmetric again
                            else -> throw AssertionError("impossible by design")
                        }
                    }
                }
            } else // scalar or ok vector
                root2
        }

        // during recompile, we may need to do inplace updates. If so, we must use the original Hop roots.
        // Todo: proper handling of dynamic recompile
        if( !doDynamicProgramRewriter ) {
            if( roots.size != roots2.size )
                throw RuntimeException("changed the number of roots during recompile from $roots to $roots2")
            roots.mapInPlaceIndexed { rootNum, root ->
                val root2 = roots2[rootNum]
                root.input.mapInPlaceIndexed { inPos, _ ->
                    val newInput = root2.input[inPos]
                    newInput.parent[newInput.parent.indexOf(root2)] = root
                    newInput
                }
                root
            }
            roots2 = roots
        }

        if (LOG.isTraceEnabled)
            LOG.trace("Spoof2Compiler created modified HOP DAG: \n" + Explain.explainHops(roots2))
        HopDagValidator.validateHopDag(roots2)

        //rewrite after applied sum-product optimizer
        ProgramRewriter(RewriteCommonSubexpressionElimination()).rewriteHopDAG(roots, ProgramRewriteStatus())
        ProgramRewriter(RewriteConstantFolding()).rewriteHopDAG(roots, ProgramRewriteStatus())
//        ProgramRewriter(RewriteAlgebraicSimplificationStatic()).rewriteHopDAG(roots, ProgramRewriteStatus())
        ProgramRewriter(RewriteCommonSubexpressionElimination()).rewriteHopDAG(roots, ProgramRewriteStatus())
        programRewriteHops(roots2, recompile, doDynamicProgramRewriter) // Todo Testing...
        //programRewriteHops(roots2, false, true)
//        if( doDynamicProgramRewriter )
//            RewriteAlgebraicSimplificationDynamic().rewriteHopDAGs(roots2, ProgramRewriteStatus())

        HopDagValidator.validateHopDag(roots2)
        if (LOG.isTraceEnabled)
            LOG.trace("Spoof2Compiler rewritten modified HOP DAG: \n" + Explain.explainHops(roots2))

        return roots2
    }

    private fun transposeRoot(root2: Hop): Hop {
        return if( root2 is DataOp && root2.isWrite ) {
            val child = root2.input[0]
            child.parent -= root2
            root2.input[0] = HopRewriteUtils.createTranspose(child)
            root2.input[0].parent += root2
            root2.refreshSizeInformation()
            root2
        } else
            HopRewriteUtils.createTranspose(root2)
    }







}